app/main.py

```diff
@@ -86,129 +86,61 @@
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
-
-    # Return battle results:
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
-
-
-print(battle(KNIGHTS))
+class BattleOfKnights:
+
+    def __init__(self,
+                 name: str,
+                 power: int,
+                 hp: int,
+                 armour: list,
+                 weapon: dict,
+                 potion: dict,
+                 protection: int = 0) -> None:
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = protection
+        self.add_protection()
+        self.add_power_from_weapon()
+        self.add_potion()
+
+    def add_protection(self) -> None:
+        for armour in self.armour:
+            self.protection += armour["protection"]
+
+    def add_power_from_weapon(self) -> None:
+        self.power += self.weapon["power"]
+
+    def add_potion(self) -> None:
+        if self.potion:
+            if "power" in self.potion["effect"]:
+                self.power += self.potion["effect"]["power"]
+
+            if "hp" in self.potion["effect"]:
+                self.hp += self.potion["effect"]["hp"]
+
+            if "protection" in self.potion["effect"]:
+                self.protection += self.potion["effect"]["protection"]
+
+    def fight(self, other: "BattleOfKnights") -> None:
+        self.hp = max(0, self.hp - other.power + self.protection)
+        other.hp = max(0, other.hp - self.power + other.protection)
+
+
+def battle(knightsconfig: dict) -> dict:
+    dict_of_knights = {}
+    for knight in knightsconfig:
+        dict_of_knights[knight] = BattleOfKnights(
+            knightsconfig[knight]["name"],
+            knightsconfig[knight]["power"],
+            knightsconfig[knight]["hp"],
+            knightsconfig[knight]["armour"],
+            knightsconfig[knight]["weapon"],
+            knightsconfig[knight]["potion"],
+        )
```

Comment:
```suggestion
        dict_of_knights[knight] = BattleOfKnights(**knightsconfig[knight])
```
Try using unpacking here.